### PR TITLE
Solution - Calculate the factorial of a non negative integer

### PR DESCRIPTION
**Pull request by AhmedKhalifa7**
---
name: Solution Review 
about: Code review checklist
---

## Behavior

Calculate the factorial of a positive number by calculating the product of all positive integers less than or equal to the number.

### Files

- [ ] The file name describes the function's behavior
- [ ] There is a module docstring in the function file
- [ ] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [ ] There is a module docstring in the tests file

### Unit Tests

- [ ] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Every unit test has
  - [ ] A helpful name
  - [ ] A clear docstring
  - [ ] Only one assertion
  - [ ] There is no logic in the unit test
- [ ] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases

### Function Docstring

- [ ] The function's behavior is described
- The function's arguments are described:
  - [ ] Type
  - [ ] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [ ] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [ ] The function's name describes it's behavior
- [ ] The function's name matches the file name
- [ ] The function has correct type annotations

### Do's

- [ ] Variable names help to understand the strategy
- [ ] Any comments are clear and describe the strategy
- [ ] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [ ] The function's strategy _is not_ described in the documentation
- [ ] Comments explain the _strategy_, **not** the _implementation_
- [ ] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [ ] The code passes the formatting checks
- [ ] The code passes all Ruff linting checks
- [ ] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [ ] Variables are named with snake_case
- [ ] Variable names are clear and helpful
- [ ] The code follows the strategy as simply as possible
- [ ] The implementation is as simple as possible given the strategy
- [ ] The code includes defensive assertions
- [ ] Defensive assertions include as little logic as possible
- [ ]

### CODE REVIEW

**Review by MohammadRAlSalloum — APPROVED** [review]
Looks great!

File: solutions/calculate_factorial.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env python3
```
Comment (RaedAlyan) [inline solutions/calculate_factorial.py:1]: are you working on Linux or Window OS?

File: solutions/calculate_factorial.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+    Returns -> int, the factorial of the number, greater than zero
+
+    Raises:
+        AssertionError: if the argument is not an integer
+        ValueError: if the argument is less than 0
+
+    >>> calculate_factorial(0)
+    1
+
```
Comment (RaedAlyan) [inline solutions/calculate_factorial.py:25]: remove this blank line

File: solutions/calculate_factorial.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    >>> calculate_factorial(2)
+    2
+
```
Comment (RaedAlyan) [inline solutions/calculate_factorial.py:28]: remove this blank line

File: solutions/calculate_factorial.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    >>> calculate_factorial(5)
+    120
+
```
Comment (RaedAlyan) [inline solutions/calculate_factorial.py:31]: the same here

File: solutions/calculate_factorial.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    >>> calculate_factorial(10)
+    3628800
+
+    >>> calculate_factorial(42)
+    1405006117752879898543142606244511569936384000000000
+
```
Comment (RaedAlyan) [inline solutions/calculate_factorial.py:37]: the same here

File: solutions/tests/test_calculate_factorial.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+    def test_two_digits_number2(self):
+        """Test the factorial of 42 is 1405006117752879898543142606244511569936384000000000"""
+        self.assertEqual(
+            calculate_factorial(42),
+            1405006117752879898543142606244511569936384000000000,
+        )
+
+    def test_three_digits_number(self):
+        """Test the factorial of 112 is 197450685722107402353682037275992488341277868034975337796656295094902858969771811440894224355027779366597957338237853638272334919686385621811850780464277094400000000000000000000000000"""
```
Comment (RaedAlyan) [inline solutions/tests/test_calculate_factorial.py:52]: this docstring isn't readable because the result is too long, please improve it

File: solutions/tests/test_calculate_factorial.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        self.assertEqual(
+            calculate_factorial(112),
+            197450685722107402353682037275992488341277868034975337796656295094902858969771811440894224355027779366597957338237853638272334919686385621811850780464277094400000000000000000000000000,
```
Comment (RaedAlyan) [inline solutions/tests/test_calculate_factorial.py:55]: the same here as well

File: solutions/calculate_factorial.py (hunk 1 above)
Comment (AhmedKhalifa7) [reply solutions/calculate_factorial.py:1]: No, honestly I don't know what it's about, I'm just adding it because Evan said so.

File: solutions/tests/test_calculate_factorial.py (hunk 6 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_calculate_factorial.py:52]: > this docstring isn't readable because the result is too long, please improve it
How? It passed the py_linting format, should it not pass it?

File: solutions/tests/test_calculate_factorial.py (hunk 7 above)
Comment (AhmedKhalifa7) [reply solutions/tests/test_calculate_factorial.py:55]: > the same here as well

How?

File: solutions/calculate_factorial.py (hunk 2 above)
Comment (AhmedKhalifa7) [reply solutions/calculate_factorial.py:25]: > remove this blank line

I put it to make the docstring more readable, do I have to remove?

File: solutions/calculate_factorial.py (hunk 3 above)
Comment (AhmedKhalifa7) [reply solutions/calculate_factorial.py:28]: > remove this blank line

I add it to make the docstring more readable, do I have to remove?

**Comment by MahmoudAlnouri** [comment]
Looks fine!

**Review by MahmoudAlnouri — COMMENTED** [review]
Looks good

File: solutions/calculate_factorial.py (hunk 4 above)
Comment (AhmedKhalifa7) [reply solutions/calculate_factorial.py:31]: > the same here

I did it to make the docstring readable, do I have to remove them?

File: solutions/calculate_factorial.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    >>> calculate_factorial(2)
+    2
+
+    >>> calculate_factorial(5)
+    120
+
+    >>> calculate_factorial(10)
+    3628800
+
```
[the remark this answers is not in the document]
Comment (AhmedKhalifa7) [reply solutions/calculate_factorial.py:34]: > the same here

I did it to make the docstring readable, do I have to remove them?

**Review by tamarasaqer — APPROVED** [review]
Some people write code, others simply craft it.
Good job.

**Review by RaedAlyan — APPROVED** [review]
This PR looks fine to me!
Good job @AhmedKhalifa7
